### DISCUSSION TITLE
How to build autocomplete for a custom search query language (DSL)

### DISCUSSION THREAD

**Discussion by KShivendu** [tree-sitter/tree-sitter#1328] [Q&A]
Hi, 
I'm working for a non-profit open-source organization [Software Heritage](https://archive.softwareheritage.org/) to build a search query language that looks like: 
```
visits > 5 and (license in [MIT, BSD] or metadata = "foo bar baz")
``` 

I'm supposed to build an autocomplete for this grammar. But I couldn't find any clear explanation about the approach that I should take to achieve this autocompletion. If you have any past experience in the field. Please comment. 

Thanks :D

**Comment by razzeee** [comment]
Shouldn't this be part of https://github.com/tree-sitter/tree-sitter/discussions/1149 ?

**Reply by KShivendu** [reply]
Hi @Razzeee, I agree that it should be part of the same discussion. But I thought that I couldn't get any response for my last comment (https://github.com/tree-sitter/tree-sitter/discussions/1149#discussioncomment-838279) so maybe creating a new thread could be helpful ( it might get noticed by someone )

I'm sorry if you felt that I'm cluttering the discussion section.

Btw, since you've read this thread, can you please suggest something that answers my question?

Regards

**Reply by razzeee** [reply]
I think your question is way to context sensitive. I basically answered your old question and would think that answer is as good as you could get with the question you've provided.

It's unclear where you want to autocomplete in that query. Do you need to query databases, online sources or just the local disk? Do you already have a working tree sitter grammar?

The grammar might only be helpful for autocomplete if there are other files that play into this (imports, other files with functions that you need to find etc.)

Ask your self where in that query suggestions would be helpful, write them down, try to figure out, where you can get the data to fill these in from.
